storage/src/journal/contiguous/variable.rs

```diff
@@ -0,0 +1,1519 @@
+    /// what SHOULD be in the offsets journal, then fixes any mismatches.
+    ///
+    /// # Returns
+    ///
+    /// Returns `(oldest_retained_pos, size)` for the contiguous journal.
+    async fn repair_journals(
+        data: &mut variable::Journal<E, V>,
+        offsets: &mut fixed::Journal<E, u32>,
+        items_per_section: u64,
+    ) -> Result<(u64, u64), Error> {
+        // === Handle empty data journal case ===
+        let items_in_last_section = match data.blobs.last_key_value() {
+            Some((last_section, _)) => {
+                let stream = data.replay(*last_section, 0, REPLAY_BUFFER_SIZE).await?;
+                futures::pin_mut!(stream);
+                let mut count = 0u64;
+                while let Some(result) = stream.next().await {
+                    result?; // Propagate replay errors (corruption, etc.)
+                    count += 1;
+                }
+                count
+            }
+            None => 0,
+        };
+
+        // Data journal is empty if there are no sections or if there is one section and it has no items.
+        // The latter should only occur if a crash occured after opening a data journal blob but
+        // before writing to it.
+        let data_empty =
+            data.blobs.is_empty() || (data.blobs.len() == 1 && items_in_last_section == 0);
+        if data_empty {
+            let size = offsets.size().await?;
+
+            if !data.blobs.is_empty() {
+                // A section exists but contains 0 items. This can happen in two cases:
+                // 1. Rewind crash: we rewound the data journal but crashed before rewinding offsets
+                // 2. First append crash: we opened the first section blob but crashed before writing to it
+                // In both cases, calculate target position from the first remaining section
+                let first_section = *data.blobs.first_key_value().unwrap().0;
+                let target_pos = first_section * items_per_section;
+
+                info!("crash repair: rewinding offsets from {size} to {target_pos}");
+                offsets.rewind(target_pos).await?;
+                offsets.sync().await?;
+                return Ok((target_pos, target_pos));
+            }
+
+            // data.blobs is empty. This can happen in two cases:
+            // 1. Rewind crash: we completely pruned the data journal but crashed before rewinding
+            //    the offsets journal.
+            // 2. The data journal was never opened.
+            if let Some(oldest) = offsets.oldest_retained_pos().await? {
+                if oldest < size {
+                    // Offsets has unpruned entries but data is gone - repair by pruning
+                    info!("crash repair: pruning offsets to {size} (prune-all crash)");
+                    offsets.prune(size).await?;
+                    offsets.sync().await?;
+                }
+            }
+
+            return Ok((size, size));
```

Comment:
If I'm not mistaken, it is possible that the `size` returned as the first item here isn't section-aligned?

I think that could cause the `position_to_section` function to work incorrectly?

Reply:
Not sure if this is a problem if the invariant "all sections other than last are full" is meant to be upheld?

Reply:
I think this could particularly affect the correctness of `position_to_section`

Reply:
https://github.com/commonwarexyz/monorepo/pull/1986

Reply:
Wait actually we don't allow pruning away a section that isn't full